### ISSUE TITLE
This typing for callbacks typed as any

### ISSUE THREAD

**Issue by tinganho** [Microsoft/TypeScript#10972]
I don't quite understand why below code is typed as `any`(I'm using nightly with `--noImplicitThis`)?
```ts
interface A {
    a: string;
}
declare function test(callback: (this: A) => void): void;
test(function() {
    this.a;// typed as any?
});
```

**Comment by sandersn** [comment]
This works for me when I paste this into a VS Code instance running on a branch that I merged from master on Thursday. Let me try the current master...do you have another repro besides this snippet?

**Comment by sandersn** [comment]
Never mind, I was actually running code from July. It's broken once I point VS to the Thursday branch.

**Comment by sandersn** [comment]
In case I want to `git bisect` later, July 6 was the time that the "works-for-me" branch was last merged from master.

**Comment by sandersn** [comment]
I suspect this PR: https://github.com/Microsoft/TypeScript/pull/9746
I'll investigate later.

**Comment by sandersn** [comment]
So, interestingly, this is a bug only in quick info and listing members. Change the callback to return `A` and you will see that errors are still caught:

```ts
interface A {
    a: string;
}
declare function noThis(callback: (this: A) => A): A;
noThis(function() { // error, 'string' is not assignable to 'A'
    return this.a + this.nope; // error, 'nope' does not exist on 'A'
});
```

I still suspect #9746.
